### ISSUE TITLE
Building essentia extractors

### ISSUE THREAD

**Issue by regorxxx** [wo80/essentia#5]
I want to build essentia to get extractors like the ones found here:
https://essentia.upf.edu/extractors/

In particular:
essentia_streaming_extractor_music.exe
essentia_streaming_key.exe

Ideally; I would like to compile 'essentia_streaming_extractor_music' with Gaia support to be able to run the models associated to mood, etc. No publicly available extractor was built with Gaia support, so there is no way to recreate what AcousticBrainz was doing locally within a plugins ecosystem (beets or foobar2000) -which is my main goal-. Python is out of this equation, since it must be self-contained, portable and easy for regular users.

If Gaia is not possible, I'm fine with tensorflow support as long as it works the same. i.e. just downloading the models, pointing the YML profile to them and done. Have not tested this, since I believe the public compiled extractors where only supposed to work with Gaia (?)

Anyway, I forked your repo and per your instructions tried to build it:
https://github.com/regorxxx/essentia/actions/runs/15421937087/workflow

Changed build type to 'Release' and added '-BUILD_EXAMPLES=ON' to cmake cmd line.

This is the latest output:
https://github.com/regorxxx/essentia/actions/runs/15422603374/job/43401432789#step:5:9

**Comment by wo80** [comment]
> Changed build type to 'Release'

That's fine once you get to a point where the workflow is working as expected. For now, switch back to `Debug` since `Release` takes a significant amount of time longer to build.

> and added '-BUILD_EXAMPLES=ON' to cmake cmd line.

CMake configuration options are prefixed with `-D`, so the correct syntax is `-DBUILD_EXAMPLES=ON` (or with a whitespace
`-D BUILD_EXAMPLES=ON`)

I saw you copied the `Upload artifact` workflow step. This won't work, since the specified path only contains the Python wheel.
You will need to change that to the path of the executables you want to upload. Those will probably reside in the folder `src/${{env.BUILD_TYPE}}` (we can sort this out once the build step is working). Please also refer to https://github.com/actions/upload-artifact for configuration of this step.

Finally, you could disable/remove the ubuntu and macos build steps to reduce noise.

---

We will see about Gaia once the above problems are sorted out. To build Essentia with SVM model support, you would need to download the required Qt dependencies and build Gaia2 inside the workflow, which might be a bit of a challenge for someone with limited experience.

**Comment by regorxxx** [comment]
> That's fine once you get to a point where the workflow is working as expected. For now, switch back to Debug since Release takes a significant amount of time longer to build.

Ok, noted.

> CMake configuration options are prefixed with -D, so the correct syntax is -DBUILD_EXAMPLES=ON (or with a whitespace
-D BUILD_EXAMPLES=ON)

Newbie error. Fixed!

> saw you copied the Upload artifact workflow step. This won't work, since the specified path only contains the Python wheel.

I supposed that, that's why I uploaded the entire folder; which is obviously suboptimal. After downloading the 4gb... turns out the executables are at:
\src\examples\Debug\

> Finally, you could disable/remove the ubuntu and macos build steps to reduce noise.

Done.

**Comment by wo80** [comment]
> But running any of the executables leads to crashes due to mising:

The workflow will use shared dependencies by default. Those are built by the ` Install dependencies` step using my `build-dependencies-msvc.bat` script. The dlls are located in `packaging\msvc\bin`, which could be added to the artifact.

If you want to build against static libraries, the batch script has an option `--static` and the CMake configuration has an option `ENABLE_STATIC_DEPENDENCIES=ON`. The `cmake-python` branch uses this, see https://github.com/wo80/essentia/blob/cmake-python/.github/workflows/build-cmake.yml#L90

> After downloading the 4gb...

To speed up the build step, you can delete unused examples from [src/examples/CMakeLists.txt](https://github.com/wo80/essentia/blob/cmake/src/examples/CMakeLists.txt). Also keep in mind that the debug binaries are significantly larger, so the release artifact will be smaller.

**Comment by regorxxx** [comment]
> Any project is linked against [VC runtime](https://learn.microsoft.com/en-us/cpp/windows/latest-supported-vc-redist?view=msvc-170) shared dlls by default (this is a dependency automatically created by using the VC compiler). It's possible to use the static libraries using the /MT[d] [compiler option](https://learn.microsoft.com/en-us/cpp/build/reference/md-mt-ld-use-run-time-library?view=msvc-170), which could be set [here](https://github.com/wo80/essentia/blob/cmake/src/CMakeLists.txt#L1) using CMakes [add_compile_options](https://cmake.org/cmake/help/latest/command/add_compile_options.html). Be aware that this flag would have to be set consistently for all dependencies, which might not be worth the trouble. Most Windows systems should have the VC runtime installed anyway.

I have all VC runtimes installed. Check the DLL names, they are pointing to non standard dlls with '_D' suffix.
https://developercommunity.visualstudio.com/t/vcruntime140-1ddll-is-missing/888909

Really beyond my current knowledge, but as far as I understand that's a problem due to using the Debug build?

> Tensorflow is distributed as a shared dll only (see batch script https://github.com/wo80/essentia/blob/cmake/packaging/build-dependencies-msvc.bat#L315). The v2.16 dll has a size of 232 MB and unfortunately newer versions got https://github.com/tensorflow/tensorflow/issues/63384#issuecomment-2482713195, approaching 1GB.
Thanks for the hint! I'm fine adding that dll file to the build as is.

**Comment by regorxxx** [comment]
> Regarding Gaia2: Qt is also distributed as shared dlls only, so if you don't want to get a dependency on those dlls (QtCore and QtConcurrent), you'd have to compile Qt from source in the workflow, following for example my instructions at https://wo80.webspace.rocks/posts/2023/01-compile-qt-win/

I would also be fine with the dlls along the executables. 

I mean if I can get rid of it, great; but as long as it works and I can provide it all as a single zip file which people can use easily, I don't mind filesize.

**Comment by regorxxx** [comment]
> Yes, makes sense that those don't get installed by default. Just switch the workflow back to Release. Make sure to also build the dependencies in release mode, like done here https://github.com/wo80/essentia/blob/cmake-python/.github/workflows/build-cmake.yml#L90

Yep, missed that part on first try; thanks! Will set it to Release instead of ${{env.BUILD_TYPE}}, since they will only work on such build type then.

![Image](https://github.com/user-attachments/assets/eb3c01bd-ece4-42bb-b7b0-30918e2969d7)

So finally got it working with tensor flow. BUT as far as I have seen, streaming_extractor_music highlevel ouput only works with gaia.
https://essentia.upf.edu/reference/std_MusicExtractor.html

So still zero idea about how to make it work with tensorflow models. Will create an issue for that. Maybe they are not even supported on the extractors... that would make gaia a requisite again.

> It's only two, rather small Qt dlls, iirc. So that would be the easiest way to go forward.

For sure.

> I might find some time to integrate the Gaia build into the batch script on the weekend. So for now, just start testing with the artifact of your workflow.

That would be really helpful. As I said, as long as tensorflow actually works, that's fine. But having Gaia, even if deprecated would be pretty great since there would be at least one working version, even if it's for legacy usage.

**Comment by wo80** [comment]
> BUT as far as I have seen, streaming_extractor_music highlevel ouput only works with gaia.

Yes, Tensorflow models can be used with 4 example programs only, see https://github.com/wo80/essentia/blob/cmake/src/examples/CMakeLists.txt#L96

I have tested `streaming_musicnn_predict` with the `feature-extractors/musicnn/msd-musicnn-1.pb` model, which worked as expected. My impression is, that the examples only cover a very small subset of what is actually possible with the models provided under https://essentia.upf.edu/models.html

**Comment by regorxxx** [comment]
So essentially, correct me if you find something more... things working are:
streaming_vggish_predict.exe -> with some vgg models

I tested it with mood_acoustic-vggish-audioset-1.pb and works

streaming_musicnn_predict.exe -> musiccnn models

But then.. as you say, msd-musicnn-1.pb works fine. But mood_happy-msd-musicnn-1.pb for ex. doesn't work.

**Comment by wo80** [comment]
Yes, I guess it's a matter of trial and error which models work with the examples, and which ones don't...

> streaming_tensorflowpredict -> No idea

Same for me.

**Comment by wo80** [comment]
> Yes, I guess it's a matter of trial and error which models work with the examples, and which ones don't...

After digging a bit deeper, I think it doesn't have to be trial and error.

If you look at the models at https://essentia.upf.edu/models.html, they provide weights and metadata. The metadata will help you understand what inputs a model expects and what outputs it provides.

First lets have a look at [msd-musicnn](https://essentia.upf.edu/models/feature-extractors/musicnn/msd-musicnn-1.json). The model has two relevant outputs:
1) `model/Sigmoid`, size 50, which are the actual predictions, corresponding to the genre classes
2) `model/dense/BiasAdd`, size 200, those are embeddings which are supposed to be used by other musicnn models

So, those embedding could for example be fed to [mood_acoustic-msd-musicnn](https://essentia.upf.edu/models/classification-heads/mood_acoustic/mood_acoustic-msd-musicnn-1.json):
The model expects an input of size 200, which corresponds to the `model/dense/BiasAdd` output of the msd-musicnn model. It has one relevant output `model/Softmax` of size 2, which is the prediction value (acoustic, not acoustic).

Now, lets take a look at [audioset-vggish](https://essentia.upf.edu/models/feature-extractors/vggish/audioset-vggish-3.json):

It has only one output `model/vggish/embeddings` of size 128, which are the embeddings. Those could for example be fed to [mood_happy-audioset-vggish](https://essentia.upf.edu/models/classification-heads/mood_happy/mood_happy-audioset-vggish-1.json), which again provides the output `model/Softmax` of size 2, which is the prediction value (happy, not happy).

So, in conclusion, while most of the models seem to expect this 2-step workflow (input -> embeddings -> prediction), the example programs do not support this. This means the only usable model for now seems to be `msd-musicnn`, which directly give predictions as output.

> I tested it with mood_acoustic-vggish-audioset-1.pb and works

This would actually contradict my findings, are you sure this model worked with the `streaming_vggish_predict.exe`

**Comment by wo80** [comment]
Ok, so you probably downloaded the model from

    https://essentia.upf.edu/models/classifiers/mood_acoustic/

and not

    https://essentia.upf.edu/models/classification-heads/mood_acoustic/

and the model file is about 280MB?

The models in `classification-heads/` are significantly smaller. So the advantage of using the embeddings is to have only one large "master" model which provides the embeddings, and smaller models that do the predictions.

This is finally beginning to make sense.

But then the models in `classifiers/` should actually all work out of the box. Can you give an example of a vggish model that didn't work for you?

**Comment by wo80** [comment]
The [build script](https://github.com/wo80/essentia/blob/cmake/packaging/build-dependencies-msvc.bat) now has an option `--with-gaia` to fetch Qt and build Gaia:

    .\packaging\build-dependencies-msvc.bat --static --with-gaia --build-type Release

A project  with releases of the latest Qt5 is available at
* https://github.com/wo80/qt-msvc-build/releases

You get a dependency on `Qt5Core.dll` and `Qt5Concurrent.dll`.

To build Essentia with support for Gaia SVM models, configure the CMake project with `USE_GAIA2=ON`.

I have only used the TensorFlow models and never tested the SVM models, so I'd appreciate some feedback, in case you're planning to try this out.

**Comment by regorxxx** [comment]
Thanks for the update, will try it and report.

**Comment by wo80** [comment]
I have updated [Gaia](https://github.com/wo80/gaia/commit/301b7c9f95c56628faa48fbfea1f8f4529c0237f) and the [dependencies build script](https://github.com/wo80/essentia/commit/e84dfb9b5246d65417d9c822a366ea417e89febc). Please try again and report back!

**Comment by wo80** [comment]
You haven't merged the changes in `build-dependencies-msvc.bat`

**Comment by regorxxx** [comment]
Ups, you are right. Github synced the wrong branch. Fixed now, another error:
https://github.com/regorxxx/essentia/actions/runs/15591019231/job/43909850969#step:5:918

**Comment by wo80** [comment]
Alright. It seems everything is built correctly, but the tests fail with `Exit code 0xc0000135`. This error means a DLL isn't found. Since everything except Qt is linked statically, we know the problem is Qt. This dependency is added by Gaia and since I'm not planning to support Gaia officially, you will have to fix it in your workflow file. Modify the test section and add the Qt binary dir to PATH:

    set PATH=%PATH%;%CD%\packaging\msvc\bin;%CD%\packaging\msvc\Qt5\bin

I think this should finally make the workflow pass. Fingers crossed!

**Comment by regorxxx** [comment]
Unless I did something wrong, nope. 
https://github.com/regorxxx/essentia/actions/runs/15592904255/job/43916170465

It gives same error, but the dll should be there (?)

**Comment by wo80** [comment]
Hey ho, I guess I got another idea. Since the test executable was failing, I assumed that the _Test_ step failed. But it's actually the _Build_ step which also seems to try to execute the test executable. So as done for the `Test` step, you'd also have to add `%CD%\packaging\msvc\Qt5\bin` to `PATH` in the build step.

**Comment by wo80** [comment]
It seems you haven't added `BUILD_TESTS=OFF` to the _Configure_ step, so your current build will also fail.

**Comment by wo80** [comment]
I have updated the batch script. After merging, you will have to make sure to specify `--with-tensorflow` to install TensorFlow (similar to Gaia), see https://github.com/wo80/essentia/commit/1781421783f4df60291c5cb9e05853c0644078fb

> Seems to run fine when I added tensorflow.dll and Qt5Core.dll. QtConcurrent.dll was not needed to execute the binaries, so I guess it's not needed despite your first warning (?)

Yes, it seems only the QtConcurrent header is needed for building. Upstream Gaia mentions it as a library dependency, see https://github.com/MTG/gaia/blob/master/wscript#L178

**Comment by regorxxx** [comment]
Have not had time to test it before. The new binaries (previous to your last comment) seem to work fine in both cases, for Tensorflow and Gaia models.

For Gaia I used these ones, 2.1 beta 5
https://essentia.upf.edu/svm_models/

So far streaming_extractor_freesound.exe and streaming_extractor_music.exe work fine. The results are similar in both. The first one is noticeably slower, and they usually differ at 3rd decimal.

![Image](https://github.com/user-attachments/assets/4f336b74-00ab-4f0f-825d-178a5efbe5aa)

Judging by the source code:
https://github.com/MTG/essentia/blob/46f8bd9f5a3029ede18dfa251ccc30a9c682cf88/src/algorithms/extractor/musicextractor.cpp#L20

As far as understand the music variant uses replay gain before analysis and the freesound one omits that step (since it's focused on sound identification?). I suppose that's the difference in float numbers seen, not really delved more into it to understand why freesound is much slower.

So far I would say this closes the issue, since now both Gaia and Tensorflow may be built for win.

**Comment by regorxxx** [comment]
> I have updated the batch script. After merging, you will have to make sure to specify `--with-tensorflow` to install TensorFlow (similar to Gaia), see [1781421](https://github.com/wo80/essentia/commit/1781421783f4df60291c5cb9e05853c0644078fb)
> 
> > Seems to run fine when I added tensorflow.dll and Qt5Core.dll. QtConcurrent.dll was not needed to execute the binaries, so I guess it's not needed despite your first warning (?)
> 
> Yes, it seems only the QtConcurrent header is needed for building. Upstream Gaia mentions it as a library dependency, see https://github.com/MTG/gaia/blob/master/wscript#L178

Not applied this yet.

Thinking about it, having a dependency on those dlls for binaries which don't use them at all seems a bit weird though. 

I mean, it's ok if streaming_extractor_music_svm.exe or streaming_tensorflowpredict.exe require dlls, but what's the point of having other 20 executables with those dependencies which are not used (?). 

In the end I will be using a mix of builds. Merging standalone binaries which can work fine without dlls along the 3 or 4 associated to tensorflow and gaia. And then share those files at the repository I linked before.

**Comment by regorxxx** [comment]
> > Thinking about it, having a dependency on those dlls for binaries which don't use them at all seems a bit weird though.
> > I mean, it's ok if streaming_extractor_music_svm.exe or streaming_tensorflowpredict.exe require dlls, but what's the point of having other 20 executables with those dependencies which are not used (?). Not that it's relevant when you share all of them together, but some plugins only require a single one (streaming_key.exe for ex.) and it's absurd requiring 2 unused dlls which have bigger size than the exe (240 Mb vs 12 Mb).
> 
> I have added a bit more context to issue [#6](https://github.com/wo80/essentia/issues/6) . In short: while it would be possible to get rid of the unwanted and unnecessary dependencies, it's not straightforward to do. Additionally, it would break compatibility with upstream.

Unless they adopt those changes I don't really think it's worth it. Have adjusted 3 actions to only attach the relevant 4/2 binaries from tensorflow/gaia, so it's just a matter of merging the 3 artifacts from these builds (being the base without any library). But all  examples are built on every run, only to be discarded later...

Being able to selectively build only those binaries would be a nice solution though.

I have seen here, that it's technically possible to specify the examples to built.
https://github.com/wo80/essentia/blob/1781421783f4df60291c5cb9e05853c0644078fb/src/wscript#L34

While automatically building all examples would still face the problem you describe on the issue, Tensorflow binaries are only built when enabling Tensorflow library, so it makes some sense that someone who wants to build with that library only needs those specific 4 binaries (not the rest)

With gaia is something similar, since only 2 binaries support it. If you want gaia, probably you only want those 2 binaries.

> It seems you haven't added BUILD_TESTS=OFF to the Configure step, so your current build will also fail.

Not sure if this issue was replied, but it was resolved and all tests work :)

**Comment by wo80** [comment]
Selecting examples is now possible by setting `BUILD_EXAMPLES` to a semi-colon list of example names:

    -DBUILD_EXAMPLES="streaming_musicnn_predict;streaming_vggish_predict"

**Comment by wo80** [comment]
Those are the FFmpeg libraries and it means that Essentia was compiled using the shared FFmpeg libs but it cannot find the dlls. Most likely a flaw introduced in https://github.com/wo80/essentia/commit/e4d33ba2dcafe7c3a574f8d8f00002e3dba1a065 . I will have a look later.

I added an option `--shared` to the batch script, so while using `--static`, you can still specify libraries that should be linked using the dlls, for example `--shared ffmpeg`. Seems to be a problem in case `--shared` isn't specified at all, so a temporary workaround would be to just pass `--shared x` to the batch script (or download the missing dlls from https://github.com/wo80/ffmpeg-audio-only/releases).

**Comment by wo80** [comment]
Should be fixed in https://github.com/wo80/essentia/commit/5263c344e19125c7d3df0797e55d5a5a461ff3af

**Comment by regorxxx** [comment]
Yes, it works now.

**Comment by daFish** [comment]
Hey @wo80 & @regorxxx,

thanks for this issue and the work you've done. I also forked essentia, used your commits and added some more stuff (e.g. building also for macOS).
You can find the diff under https://github.com/wo80/essentia/compare/cmake...fmdb:essentia:cmake

I'm still trying to get everything running regarding Tensorflow from https://essentia.upf.edu/models.html & https://essentia.upf.edu/models/
